include/swift/AST/SourceFile.h

```diff
@@ -117,6 +117,13 @@ class SourceFile final : public FileUnit {
   /// The scope map that describes this source file.
   std::unique_ptr<ASTScope> Scope;
 
+  /// The set of validated opaque return type decls in the source file.
+  llvm::SmallVector<OpaqueTypeDecl *, 4> OpaqueReturnTypes;
+  llvm::StringMap<OpaqueTypeDecl *> ValidatedOpaqueReturnTypes;
+  /// The set of parsed decls with opaque return types that have not yet
+  /// been validated.
+  llvm::SetVector<ValueDecl *> UnvalidatedDeclsWithOpaqueReturnTypes;
```

Comment:
Do we still need the separate validated/unvalidated lists after this change?

Reply:
I think so. The unvalidated list contains all decls that will own an opaque return decl (we haven't synthesized the opaque return decl yet). The validated map maps mangled names to opaque type decls.

Reply:
Hm, I suppose even with requestification this might be the best way of caching the work we've already done without forcing us to eagerly find and make a mangled name for all decls with opaque types up front.

Reply:
Yeah, at some point we have to go and eagerly realize all the opaque types in a source file, either for lookup or IRGen.